### ISSUE TITLE
Faster runs using `extensionDevelopmentPath` CLI flag

### ISSUE THREAD

**Issue by timvahlbrock** [redhat-developer/vscode-extension-tester#430]
VSCode has a `extensionDevelopmentPath` cli flag, which loads the unpacked extension from the handed path. This can be especially useful during development of the extension, because the exension can be compiled in watch mode and doesn't need to be packaged everytime.

I've prepared this on [my fork](https://github.com/timvahlbrock/vscode-extension-tester), however there is currently [an issue](https://bugs.chromium.org/p/chromedriver/issues/detail?id=3985) with the Windows Chrome driver, which leads to the flag being handed to VSCode in all lower case, leaving it unconsidered. Opening this issue to check if there is any interest in supporting something like a feature like this in this project.

**Comment by christian-bromann** [comment]
Good catch on the flag issue, I was debugging in webdriverio-community/wdio-vscode-service#4 why test extensions aren't loaded. Therefor I believe bundling it with `vsce` is currently the only way managing it in Windows.

**Comment by timvahlbrock** [comment]
I thought about using a Proxy-Executable for VSCode, which would just correct the malformed argument and then pass everything on to the original VSCode.exe. Maybe this can even be done using a nodejs script? I'm not quite sure though how well this can get integrated into this module, probably needs some modifications here to call the proxy script instead of the real one.

**Comment by timvahlbrock** [comment]
>I thought about using a Proxy-Executable for VSCode, which would just correct the malformed argument and then pass everything on to the original VSCode.exe.

Gave a quick shot at this using a node script, but had no success. Looks like the chrome driver detects that the handed executable (node) is not a chrome binary.

**Comment by timvahlbrock** [comment]
So I pushed my current state, but it doesn't quite work so far. I'm getting the following error.

```
WebDriverError: unknown error: Chrome failed to start: crashed.
  (unknown error: DevToolsActivePort file doesn't exist)
  (The process started from chrome location C:\Program Files\nodejs\node.exe is no longer running, so ChromeDriver is assuming that Chrome has crashed.)
```

If I'm setting breakpoints in convertArgs.js they are never hit, but that might be caused by the script being invoked by the selenium server, and not the client. The script works if I call it manually.

**Comment by christian-bromann** [comment]
@timvahlbrock Chromedriver starts Chrome with a parameter to set the chrome devtools port to `0` which causes Chrome to start it on a random port and propagate that random port through stdout, e.g.:

```sh
❯ /Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --remote-debugging-port=0

DevTools listening on ws://127.0.0.1:56524/devtools/browser/785487c8-2cdb-4fbe-88af-3d73802c3d5e
```

If you pipe the stdout of the Chrome process into the stdout of the Node process it should work.

**Comment by timvahlbrock** [comment]
But I'm doing that already:

https://github.com/timvahlbrock/vscode-extension-tester/blob/c89962e69898158cbfbdc101ffe1b52746aa31fe/src/convertArgs.ts#L10

**Comment by christian-bromann** [comment]
mhm 🤔  I think it is more an issue of permissions? I am currently testing this in the WebdriverIO service (https://github.com/webdriverio-community/wdio-vscode-service/pull/9) I mentioned above.

Running with these capabilities:

```
[0-0] {
[0-0]     "binary": "D:\\a\\wdio-vscode-service\\wdio-vscode-service\\.wdio-vscode-service\\vscode-win32-archive-1.65.2\\Code.exe",
[0-0]     "args": [
[0-0]         "--no-sandbox",
[0-0]         "--disable-updates",
[0-0]         "--skip-welcome",
[0-0]         "--skip-release-notes",
[0-0]         "--disable-workspace-trust",
[0-0]         "--disable-extensions",
[0-0]         "--extensionDevelopmentPath=D:\\a\\wdio-vscode-service\\wdio-vscode-service\\test\\extension",
[0-0]         "--user-data-dir=C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmp-5132-zM24WJBhRtxP\\settings",
[0-0]         "--extensions-dir=C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmp-5132-zM24WJBhRtxP\\extensions",
[0-0]         "--folder-uri=D:\\a\\wdio-vscode-service\\wdio-vscode-service",
[0-0]         "--file-uri=D:\\a\\wdio-vscode-service\\wdio-vscode-service\\README.md",
[0-0]         "--verbose",
[0-0]         "--logExtensionHostCommunication"
[0-0]     ],
[0-0]     "windowTypes": [
[0-0]         "webview"
[0-0]     ]
[0-0] }
```

It seems it doesn't even find the folder:

![statusbar - can get current encoding](https://user-images.githubusercontent.com/731337/160811181-ecc4413e-54ca-4065-a0bc-9d19c2e97e1e.png)

If I replace them with forward slashes, e.g.:
```
[0-0] {
[0-0]     "binary": "D:\\a\\wdio-vscode-service\\wdio-vscode-service\\.wdio-vscode-service\\vscode-win32-archive-1.65.2\\Code.exe",
[0-0]     "args": [
[0-0]         "--no-sandbox",
[0-0]         "--disable-updates",
[0-0]         "--skip-welcome",
[0-0]         "--skip-release-notes",
[0-0]         "--disable-workspace-trust",
[0-0]         "--disable-extensions",
[0-0]         "--extensionDevelopmentPath=D:/a/wdio-vscode-service/wdio-vscode-service/test/extension",
[0-0]         "--user-data-dir=C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmp-11[80](https://github.com/webdriverio-community/wdio-vscode-service/runs/5752109337?check_suite_focus=true#step:6:80)-0V6c2WejZveZ\\settings",
[0-0]         "--extensions-dir=C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmp-1180-0V6c2WejZveZ\\extensions",
[0-0]         "--folder-uri=D:/a/wdio-vscode-service/wdio-vscode-service",
[0-0]         "--file-uri=D:/a/wdio-vscode-service/wdio-vscode-service/README.md",
[0-0]         "--verbose",
[0-0]         "--logExtensionHostCommunication"
[0-0]     ],
[0-0]     "windowTypes": [
[0-0]         "webview"
[0-0]     ]
[0-0] }
```
I get this:

![statusbar - can receive all items](https://user-images.githubusercontent.com/731337/160811316-262db82b-e265-470a-9870-8b5b1e5e5e43.png)

Which seems to look better but also doesn't open the workspace. While this experiment is not about the `extensionDevelopmentPath` it shows that there might be some weirdness with permissions?

**Comment by timvahlbrock** [comment]
I've just tried to open folder with VS Code, I don't have permissions for. VS Code opens it as if it was a non-existing file (yes, the directory). What I find a little irritating in your screenshots: On your first one, VS Code shows the branch name, so it seems to be able to read at least some kind of data from FS. I guess the yellow exclamation mark doesn't provide any useful information on hover?

**Comment by christian-bromann** [comment]
Ok, I was able to figure it out: adding a schema to `--folder-uri` and `--file-uri`, e.g. `--folder-uri=file:C:/...` works but doesn't have an effect on the behavior of `extensionDevelopmentPath`.

**Comment by christian-bromann** [comment]
@timvahlbrock good news! I found [a solution](https://github.com/webdriverio-community/wdio-vscode-service/pull/4) for this problem using the fake binary method. In my Chrome capabilities I make sure to decamelize all parameters and in the fake binary I camelcase them again (basically passing them along as camelcased and decamelized version). The only disadvantage this solution has is that Windows doesn't like to run Node.js modules as binaries through Chromedriver. My workaround currently is to package it into a `.exe` using [pkg](https://www.npmjs.com/package/pkg). This obviously increases the package size but for now I am not so worried about that.

Feel free to give it a try a let me know.

**Comment by timvahlbrock** [comment]
>The only disadvantage this solution has is that Windows doesn't like to run Node.js modules as binaries through Chromedriver. 

I suspected something like this, which is why I wasn't sure whether it would work using a node script. But very nice that you have found a solution. Nevertheless I have switched back to using the integration-test framework provided by VS Code, because the tests not only took too long to run but also failed every third or fourth time because of some strange timing issues in the VS Code UI (especially: "Initializing JS/TS language features" and "Activating Extensions" have disappeared from the status bar, but VS Code wasn't quite ready to ask extensions for Code Actions yet). I saw that you are proxying the VS Code API in your project, so maybe that would be an opportunity to get things like this out, so I might check again on wdio-vscode-service for my next extension.

As there doesn't seem to be any interest in this feature from the maintainers of this repository however I'm closing this issue for now.
